### PR TITLE
PROTON-2332: [C++] Allow access to link properties

### PR DESCRIPTION
**Pull request by pjfawcett**
In C++ API, allow setting of link properties via sender and receiver options
Allow query of link properties set by remote end
Add tests

### CODE REVIEW

**Review by astitcher — CHANGES_REQUESTED** [review]

File: cpp/src/link_test.cpp [hunk 1]
```diff
@@ -45,8 +55,107 @@ int test_link_name()
...
+        : done(false),
+          had_receiver(false),
+          had_sender(false),
+          receiver_options(r_opts),
+          sender_options(s_opts)
+    {}
+
+    void on_container_start(proton::container &c) PN_CPP_OVERRIDE {
+        listener = c.listen("//:5678"); // Dynamic port doesn't work on MAC build - presumably not allocated until later
+        connection = c.connect("//:5678");
```
[the remark this answers is not in the document]
Comment (pjfawcett) [reply cpp/src/link_test.cpp:90]: Will do

File: cpp/src/link_test.cpp [hunk 2]
```diff
@@ -45,8 +55,107 @@ int test_link_name()
     return 0;
 }
 
+/*
+ * Want to test remote properties so need to run a container and make loopback links.
+ * Basing these classes on those used in container_test.cpp but reducing the checks on
+ * container logic as that is not the subject of these tests.
+ */
```
[the remark this answers is not in the document]
Comment (pjfawcett) [reply cpp/src/link_test.cpp:62]: I vacillated about which source file was the best home for these tests.  Happy to take your guidance and move them to the container tests.

[the remark this answers is not in the document]
Comment (pjfawcett) [reply cpp/src/link_test.cpp]: @astitcher are you happy with the updated tests?

**Review by astitcher — APPROVED** [review]
This looks good - could you create a covering JIRA; squash into a single commit; then edit the commit message to include the JIRA reference (see most of the other commit messages). (This allows us to put together the release notes and to link commits to the issue tracking system)
Then I'll be happy to merge it.

**Comment by pjfawcett** [comment]
Raised separate, sub-task, JIRA PROTON-2332 to cover just the work on link properties. 
Squashed commits and re-worded commit message to standard format

**Comment by pjfawcett** [comment]
@astitcher can this be merged now?

**Pull request merged**